### DISCUSSION TITLE
Add more layouts to the webui

### DISCUSSION THREAD

**Discussion by cristobaltapia** [dariogoetz/keyboard_layout_optimizer#67] [General]
Hi @dariogoetz ,

I noted that not all of the layouts defined in the `config/keyboard` directory are present in the webui. Is there a reason for this? It would be nice to send results from my layout to your webui and see what results people get.

**Comment by dariogoetz** [comment]
Hi Cristobal,

some of the config files won't work very well as they do not support German Umlauts by default (e.g. the qwerty ones and the chocolad one). The default evaluation metrics, however, do assume that they are defined (for the `similar_letter_groups` metric). This means that such configs would lead to errors in the webui and I therefore do not consider them there.

As for your `lily58` config, I have added it to the web-evaluation (github.io) as well as to the database web-ui (fly.dev).

As I had a closer look at your config (I didn't have the time before), I noticed some things that you might consider changing:
In the `key_costs`, the cost values for the keys representing `,.` and the numbers are higher than the sum of costs for the modifier for layer 4 (which also contains these symbols) and the corresponding keys on layer 4. This means that the evaluator/optimizer will assume that those symbols will be taken from layer 4 and not the base layer.
While I understand that for the numbers (the number row is very high up), I am not sure if this is realistic for `.` and `,`. Moreover, the `0` will be taken on layer 4 of the left-middle thumb key (depicted with a `♕` in the layout plots). Is that intended?

The effect in the evaluation is that there are very many finger repeats on the thumb with the layer 4 modifier and the space keys, both being on the left thumb. A considerable chunk comes from the bigram `. ` (period - space), which then after considering modifiers includes a bigram `⇚ ` (mod4 - space) on the thumb.

**Comment by cristobaltapia** [comment]
Hi Dario, 

thanks for your answer and for the explanation regarding the webui and (even more important) about the configuration. I will take a look at those costs, as I didn't understand the details of those variables at the beginning. I am pretty new to this whole layout optimization thing (although I have been wanting to start for a long while now) and I just got my keyboard as a present last weekend, so I am learning :)

**Comment by cristobaltapia** [comment]
Just to be sure:

- The symbols `⇧` and `⇗` are `Shift` (also layer 2)
- The symbols `⇩` and `⇘` are `Mod3` (layer 3)
- The symbols `⇚` and `⇙` are `Mod4` (layer 4)

Then we define their working in the `modifiers` section. But how does this mapping work? I am guessing that the first definition corresponds to layer 2, the second to layer three and so on?

I am getting a better picture now of this works. The values used now were mostly copied from the ortho layout, but now I think that I can do it a bit better.

The `0` in the left thumb is incorrect. I corrected that now.

**Reply by dariogoetz** [reply]
I merged your changes and additionally changed the thumb line break to the correct symbol (\n).

**Reply by cristobaltapia** [reply]
Thanks! Yeah, I was actually not sure about how I was supposed to write that (and tab too). It makes sense.

**Comment by cristobaltapia** [comment]
May I ask something a bit off-topic? How is your experience with programming with the bone layout? Did you use any specific program to learn the new layout (or noted)?

**Reply by dariogoetz** [reply]
My experience with the "bone" layout is rather limited compared to the "neo2", "mine" and "noted" layouts. I had slightly more issues with getting used to it compared to the other two. For neo2 and mine I used ktouch in the beginning if I recall correctly. Later I used them productively everywhere and when I had time, I did whatever was fun to do: monkeytype, typeracer, keybr, and typingclub (all of them are online-resources).

**Reply by cristobaltapia** [reply]
Thanks for those ideas! I was thinking about ktouch too, but wasn't sure how good it was. I actually was looking into it earlier today and discovered that it is possible to define new keyboard layouts (e.g. for the Lily58). I guess I'll be spending some hours configuring that this week and hopefully then finally start training :P
